Cython/Compiler/Nodes.py

```diff
@@ -10472,8 +10508,39 @@ def generate_loop(self, code, fmt_dict):
                 code.putln("#if 0")
             else:
                 code.putln("#ifdef _OPENMP")
+            with_gil_deadlock_avoidance_point = code.insertion_point()
             code.put("#pragma omp for")
 
+        if self.with_gil:
+            # Any firstprivate creates a barrier at least on GCC (and thus
+            # a deadlock if we're using the GIL). And there's also an implicit
```

Comment:
Have you examined whether this would also be relevant if someone's using critical sections inside an openmp loop? If so, should we document that?

Reply:
I'm assuming you mean `with cython.critical_section`? (I ask because OpenMP has a "critical_section" feature that we use internally to guard certain small blocks... I think we use that safely though).

I don't think that's `cython.critical_section` should be an issue here. The barrier is just that it waits for all threads to be ready to go before the absolutely first iteration.  So it's just trying not to hold the GIL when that happens. There shouldn't be an issue with anyone using critical sections or other kinds of locks within the loop itself.